### PR TITLE
feat: alloy-dyn-contract

### PR DESCRIPTION
**Pull request by onbjerg**
## Motivation

We want something like `BaseContract`/`ContractInstance` in ethers-rs for dynamic ABI in alloy.

## Solution

Port over what we can of those existing APIs to alloy (i.e. really only `call`) and simplify the `CallBuilder` a bit

In ethers there are 2 call builders, a main one, `FunctionCall`, and a secondary one you can get from `FunctionCall` (called `RawCall`) which you can use to set state overrides. I inlined the state override API into `CallBuilder` instead, and opted to only add the 3rd parameter for `eth_call` if state overrides are set, as this is not supported by all nodes.

I'm not sure if there is any risk in sending a third "bogus" parameter in `eth_call`, but if so, maybe the API in this PR could be updated to split out the `eth_call` + state override to a different function in the provider.

This also does not include sending transactions (that part is a `todo!()`) because that flow is not really done yet in the provider or the signer.

A separate pr for "static" contracts should be opened

This PR does also not differentiate between legacy/EIP-1559 tx's as we don't have a "typed tx" construct in Alloy yet

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/alloy-dyn-contract/src/call.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    pub const fn gas(mut self, gas: U256) -> Self {
+        self.request.gas = Some(gas);
+        self
+    }
+
+    /// Sets the `gas_price` field in the transaction to the provided value
+    /// If the internal transaction is an EIP-1559 one, then it sets both
+    /// `max_fee_per_gas` and `max_priority_fee_per_gas` to the same value
+    pub const fn gas_price(mut self, gas_price: U256) -> Self {
+        // todo: differentiate between eip-1559 and legacy once we have typedtx
```
Comment (Evalir) [inline crates/alloy-dyn-contract/src/call.rs:62]: I'd say let's default to legacy txs while we don't have typedtx—it's what we use in tests IIRC

File: crates/alloy-dyn-contract/src/call.rs [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+///
+/// The builder can be `.await`ed directly which is equivalent to invoking [`CallBuilder::call`].
+///
+/// # Note
+///
+/// Sets the [state overrides](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set) for `eth_call`, but this is not supported by all clients.
+pub struct CallBuilder<P> {
+    // todo: this will not work with `send_transaction` and does not differentiate between EIP-1559
+    // and legacy tx
+    request: CallRequest,
```
Comment (Evalir) [inline crates/alloy-dyn-contract/src/call.rs:23]: I think having access to the CallRequest would be ideal. This way we can get around not having `send_transaction` on the provider by transferring the fields manually to a signable/rlp-able tx type from alloy consensus e.g `TxLegacy`

File: crates/alloy-dyn-contract/src/call.rs [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+impl<P> Clone for CallBuilder<P>
+where
+    P: Clone,
+{
+    fn clone(&self) -> Self {
+        CallBuilder {
+            request: self.request.clone(),
+            function: self.function.clone(),
+            block: self.block,
```
Comment (Evalir) [inline crates/alloy-dyn-contract/src/call.rs:150]: why are we not copying the block? can't we derive this automatically and put a `P: Clone` boundary?

File: crates/alloy-dyn-contract/src/interface.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+use crate::{
+    error::{Error, Result},
+    instance::ContractInstance,
+};
+use alloy_dyn_abi::{DynSolValue, FunctionExt, JsonAbiExt};
+use alloy_json_abi::{Function, JsonAbi};
+use alloy_primitives::{Address, Selector};
+
+/// A smart contract interface.
+#[derive(Debug, Clone)]
```
Comment (Evalir) [inline crates/alloy-dyn-contract/src/interface.rs:12]: Could we add a bit more docs on why it's necessary/more ergonomic to have this type vs extending `JsonAbi`?

File: crates/alloy-dyn-contract/src/interface.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    functions: HashMap<Selector, (String, usize)>,
+}
+
+impl From<JsonAbi> for Interface {
+    fn from(abi: JsonAbi) -> Self {
+        Self { abi, functions: Default::default() }
+    }
+}
+
+// TODO: events/errors
```
Comment (Evalir) [inline crates/alloy-dyn-contract/src/interface.rs:24]: just flagging :D

File: crates/providers/src/provider.rs [hunk 6]
```diff
@@ -359,8 +368,24 @@ impl<T: Transport + Clone + Send + Sync> TempProvider for Provider<T> {
...
+    /// Not all client implementations support state overrides.
+    async fn call(
+        &self,
+        tx: CallRequest,
+        block: Option<BlockId>,
+        state: Option<StateOverride>,
+    ) -> TransportResult<Bytes> {
+        // todo: if we just set `None` for the 3rd parameter, will this be ok for clients who do not
+        // expect a third parameter? if not, should we split this into `call` and
+        // `call_with_overrides`?
```
Comment (Evalir) [inline crates/providers/src/provider.rs:383]: I think separating both is okay. Frankly, not many people use call with overrides and having to set none becomes a chore

**Review by Evalir — COMMENTED** [review]
looks great, some comments

File: crates/alloy-dyn-contract/src/call.rs [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    /// Sets the `gas_price` field in the transaction to the provided value
+    /// If the internal transaction is an EIP-1559 one, then it sets both
+    /// `max_fee_per_gas` and `max_priority_fee_per_gas` to the same value
+    pub const fn gas_price(mut self, gas_price: U256) -> Self {
+        // todo: differentiate between eip-1559 and legacy once we have typedtx
+        self.request.gas_price = Some(gas_price);
+        self.request.max_fee_per_gas = Some(gas_price);
+        self.request.max_priority_fee_per_gas = Some(gas_price);
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/call.rs:65]: hmm, I guess eip-1559 for now is okay then

File: crates/providers/src/provider.rs [hunk 8]
```diff
@@ -133,7 +133,16 @@ pub trait TempProvider: Send + Sync {
...
+    ///
+    /// # Note
+    ///
+    /// Not all client implementations support state overrides.
+    async fn call(
+        &self,
+        tx: CallRequest,
+        block: Option<BlockId>,
+        state: Option<StateOverride>,
+    ) -> TransportResult<Bytes>;
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/providers/src/provider.rs:145]: sg to me - was not sure what the best approach was so open to either

File: crates/alloy-dyn-contract/src/call.rs (hunk 3 above)
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/call.rs:150]: i don't want to force it to be `P: Clone`, so `Clone` is only impld for `CallBuilder` if `P: Clone`. i think that's ok

the block is `Copy`, so we don't `.clone()` or clippy will shoot us

File: crates/alloy-dyn-contract/src/interface.rs (hunk 5 above)
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/interface.rs:24]: yea we don't have the facilities rn for events so it's more of a follow up to do

Comment (Evalir) [reply crates/alloy-dyn-contract/src/interface.rs:24]: sweet np, let's open an issue for this and roll

File: crates/providers/src/provider.rs (hunk 6 above)
Comment (gakonst) [reply crates/providers/src/provider.rs:383]: Separate IMO yep

File: crates/alloy-dyn-contract/src/instance.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        this
+    }
+}
+
+impl<P: TempProvider + Clone> ContractInstance<P> {
+    /// Returns a transaction builder for the provided function name.
+    /// If there are  multiple functions with the same name due to overloading, consider using
+    /// the [`ContractInstance::function_from_selector`] method instead, since this will use the
+    /// first match.
+    pub fn function(&self, name: &str, args: &[DynSolValue]) -> Result<CallBuilder<P>> {
```
Comment (gakonst) [inline crates/alloy-dyn-contract/src/instance.rs:65]: I would love if we had a great UX about this. Right now it'd require the user to manually do `DynSolValue::X` for each argument. The way we did it with the generics `Tokenize` in Ethers-rs was nice, but understand if it's a PITA to work with.

cc @DaniPopes

File: crates/alloy-dyn-contract/src/call.rs [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    /// If executed on a mutating smart contract function, it will do a "dry run" of the call
+    /// and return the return type of the transaction without mutating the state
+    ///
+    /// # Note
+    ///
+    /// This function _does not_ send a transaction from your account
+    pub async fn call(&self) -> Result<Vec<DynSolValue>> {
+        let bytes = self
+            .provider
+            .call(self.request.clone(), self.block, self.state.clone())
```
Comment (gakonst) [inline crates/alloy-dyn-contract/src/call.rs:126]: @DaniPopes  @mattsse is there a world where we can get rid of these clones and use refs under the hood

File: crates/alloy-dyn-contract/src/call.rs (hunk 1 above)
Comment (gakonst) [reply crates/alloy-dyn-contract/src/call.rs:62]: Let's default to 1559, no?

**Review by gakonst — COMMENTED** [review]
nice

File: crates/alloy-dyn-contract/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "alloy-dyn-contract"
```
Comment (DaniPopes) [inline crates/alloy-dyn-contract/Cargo.toml:2]: needs description to be the same as readme

File: crates/alloy-dyn-contract/src/instance.rs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+    P: Clone,
+{
+    fn clone(&self) -> Self {
+        ContractInstance {
+            address: self.address,
+            provider: self.provider.clone(),
+            interface: self.interface.clone(),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/instance.rs:94]: we can't unless we require that `P: Clone` which we talked about last sync that we do not want to enforce

File: crates/alloy-dyn-contract/src/instance.rs (hunk 9 above)
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/instance.rs:65]: I tried to convey that at last sync but there did not seem to be a lot of agreement

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/alloy-dyn-contract/src/call.rs [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    Self: 'static,
+    P: TempProvider,
+{
+    type Output = Result<Vec<DynSolValue>>;
+
+    #[cfg(target_arch = "wasm32")]
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output>>>;
+
+    #[cfg(not(target_arch = "wasm32"))]
+    type IntoFuture = Pin<Box<dyn Future<Output = Self::Output> + Send>>;
```
Comment (DaniPopes) [inline crates/alloy-dyn-contract/src/call.rs:171]: Could `impl Future` work here?

File: crates/alloy-dyn-contract/src/error.rs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::UnknownFunction(name) => {
+                write!(f, "unknown function: function {name} does not exist",)
+            }
+            Self::UnknownSelector(selector) => {
+                write!(
+                    f,
+                    "unknown function: function with selector {} does not exist",
+                    hex::encode(selector)
```
Comment (DaniPopes) [inline crates/alloy-dyn-contract/src/error.rs:55]:
```suggestion
                    "unknown function: function with selector {selector} does not exist"
```

File: crates/alloy-dyn-contract/src/call.rs (hunk 3 above)
Comment (DaniPopes) [reply crates/alloy-dyn-contract/src/call.rs:150]: Sorry I don't really understand the reasoning here and below of not deriving? It's the exact same as the compiler derived implementation.

File: crates/alloy-dyn-contract/src/interface.rs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        data: &[u8],
+        validate: bool,
+    ) -> Result<Vec<DynSolValue>> {
+        self.get_from_selector(selector)?.abi_decode_output(data, validate).map_err(Into::into)
+    }
+
+    /// Returns a reference to the contract's ABI
+    pub const fn abi(&self) -> &JsonAbi {
+        &self.abi
+    }
```
Comment (DaniPopes) [inline crates/alloy-dyn-contract/src/interface.rs:103]: + `fn into_abi(self)`

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/alloy-dyn-contract/src/call.rs (hunk 3 above)
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/call.rs:150]: I see it works now, earlier I tried to derive w/o requiring `P: Clone` but it complained

File: crates/alloy-dyn-contract/src/call.rs (hunk 13 above)
Comment (onbjerg) [reply crates/alloy-dyn-contract/src/call.rs:171]: that feature is still unstable unfortunately

File: crates/dyn-contract/src/call.rs [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+/// Sets the [state overrides](https://geth.ethereum.org/docs/rpc/ns-eth#3-object---state-override-set) for `eth_call`, but this is not supported by all clients.
+#[derive(Clone)]
+pub struct CallBuilder<P> {
+    // todo: this will not work with `send_transaction` and does not differentiate between EIP-1559
+    // and legacy tx
+    request: CallRequest,
+    block: Option<BlockId>,
+    state: Option<StateOverride>,
+    provider: P,
+    // todo: only used to decode - should it be some type D to dedupe with `sol!` contracts?
```
Comment (DaniPopes) [inline crates/dyn-contract/src/call.rs:28]: Will need a follow-up for this so we can also use it in sol!

**Review by DaniPopes — APPROVED** [review]
lgtm

**Comment by onbjerg** [comment]
will merge as is and we can make improvements on top, we're prob bound to have missed something from the old api that someone has a use case for

**Pull request merged**